src/base/dense/owl_base_dense_ndarray_generic.ml

```diff
@@ -763,6 +763,50 @@ let max' varr = (_fold_left (Pervasives.max) Pervasives.min_float varr)
 let sum' varr = (_fold_left (+.) 0. varr)
 
 
+(** Folding along a specified axis, aka reduction.
+    m: number of slices; n: x's slice size; o: x' strides, also y's slice size;
+    x: source; y: shape of destination. Note that o <= n. *)
+let owl_sum_along m n o x ys =
```

Comment:
Folding along specified axes is generic operation on Ndarray. You should pass in higher-order function to make this operator generic to other number types (such as complex) and different operations such as (such as min, max, sub, mul, and etc.).

I am also curious why you don't reuse the the existing sum operation, is it because it is too slow for calling multiple times if you need to reduce on multiple axes.

Reply:
Yes, one of the concerns is speed. I saw the comment of the existing sum operation about its possible efficiency issue. Calling it multiple times might be quite slow. 

The other is that the `sum` in base lib reduce the input data by one dimension, and I'm sure if that's the intended behaviour of this function in base. For example:
```ocaml
module N = Owl_base_dense_ndarray.S
let x1 = N.sequential [|2;3;4|]
let y1 = N.sum ~axis:2 x1
let s1 = N.shape y1 (* s1 = [|2;3|] *)

module N = Owl_dense_ndarray.S
let x2 = N.sequential [|2;3;4|]
let y2 = N.sum ~axis:2 x2
let s2 = N.shape y2 (* s2 = [|2;3;1|] *)
```

Considering both factors, I decide to follow the implementation in `owl_dense_ndarray_generic.ml` using `owl_sum_along`.

Reply:
I see. Yes, folding along one dimension is common use case. OK, then please try to make the function generic, refer to owl_base_dense_common.ml

Ping me on skype if you need help on this.